### ISSUE TITLE
Synchronise version numbering of  `ModelicaServices` and `Complex` with MSL

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#811]
**Modified by dietmarw on 10 Sep 2012 15:53 UTC**
In order to avoid any confusion the packages
 * `ModelicaServices`
 * `Complex`

should get the same version number as the MSL upon release, e.g., `version="3.2.1"`. That way it is far easier to keep track of which version was shipped with which. At the same time tool vendors can apply their own sub-versioning of the versions shipped with the MSL in order to make customisations.

One way would be calling those versions (following the http://semver.org scheme) then 
`version="3.2.1+<VendorName>.<number>"`
But that is just a suggestion anyway.

----

**Reported by dietmarw on 10 Sep 2012 15:47 UTC**
In order to avoid any confusion the packages
 * `ModelicaServices`
 * `Complex`

should get the same version number as the MSL upon release, e.g., `version=3.2.1`. That way it is far easier to keep track of which version was shipped with which. At the same time tool vendors can apply their own sub-versioning of the versions shipped with the MSL in order to make customisations.

One way would be calling those versions (following the http://semver.org scheme) then 
`version=3.2.1+<VendorName>.<number>`
But that is just a suggestion anyway.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/811

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 10 Sep 2012 15:47 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 10 Sep 2012 15:53 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 11 Sep 2012 05:13 UTC**
Regarding the subversioning, we already have this in Modelica, right?
`ModelicaServices 3.2.1 XXXVisualization`

And keep ModelicaTest in line with the MSL, too.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 11 Sep 2012 08:19 UTC**
I don't think using version "3.2.1 <VendorName>..." for ModelicaServices will work directly; and would propose to instead have VendorName as part of the build-description or something similar.

First according to the spec <VendorName> is seen as similar to a beta-version (17.8.1). That does not seem right, and might lead to some minor issues.

The main reason is, however, that when someone uses ModelicaServices the VendorName would be coded into the uses-annotation of the library using ModelicaServices. That will make it harder to make tool-neutral libraries - yes they can be changed, but it will be more work.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 20 Sep 2012 14:02 UTC**
OK I forgot about the already existing sub-versioning. But anyway the sub-versioning should not be an issue for the MSL developers. I would therefore update all version numbers now to be in synch with the MSL version number and tool vendors then can decide how they want to mark their variant of ModelicaServices, either by `3.2.1 xxx` or if they don't like the pre-release smell of this they can use `3.2.1.x`.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 4 Oct 2012 09:37 UTC**
It seems that there is no good solution yet. The key issue seems to be that the whole version number is included in the "uses" annotation and then a version numbers such as "3.2.1 Dymola" in the uses annotation is bad: If different Modelica tools are used by the same user model then storing a user model with, say Dymola, and opening this model with SimulationX should give an error (or at least a warning), since SimulationX does not have a ModelicaServices package "3.2.1 Dymola", but a ModelicaServices package "3.2.1 SimulationX".

Having the version number in the forth digit does also not help, since the tools might most likely use different numbers.

Hans proposal to use the build information seems to be the only short term solution. This means that one has to coordinate between the tool vendor (whenever, a tool vendors makes its own implementation, then someone should give him a unique build-number; maybe have simply a text-file on the svn where the current associations of build-numbers to tools are defined).

As a long term solution, we probably need an additional annotation, such as 

```
versionBuildTool = "Dymola"
versionBuild = 2

```
The meaning is that this is a special implementation for tool Dymola and it is the second build. If there is quick agreement, we could even try to include such a definition already in Modelica 3.2 revision 2. I have to say that this is my preferred solution.

In the meantime, we could start with version number 3.2.1 (for ModelicaServices) and distinguish the three different variants currently on the svn server by the build number (0,1,2).

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 4 Oct 2012 09:51 UTC**
A better name might be:

```
versionBuildTarget = "Dymola"
versionBuild = 2
```

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 4 Oct 2012 12:37 UTC**
Partially fixed in [​ 3c0057c98c16b7688cf857040b7f5b8cee670109](https://trac.modelica.org/Modelica/changeset/5569)   by changing the version numbers of the ModelicaServices variants to 3.2.1 and by using different build numbers for the three variants.

I am not sure whether package Complex should have the same strategy to follow the version number of MSL. The reason is that it looks odd that for every new MSL version the version number of Complex is changed, although, there is no change in the package.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 4 Oct 2012 14:36 UTC**
> In the meantime, we could start with version number 3.2.1 (for ModelicaServices) and distinguish the three different variants currently on the svn server by the build number (0,1,2).

Martin, there is no `versionBuild=0` for **released** versions. `0` was just a placeholder until we have a release.

As for Complex, I think it's easier to remember that MSL 3.2.1 comes with Complex 3.2.1 as to find out which version of Complex was shipped with which version of MSL (even if there is no functional change). On the contrary it kind of reflects somebody has actually looked at the Complex library and "signed" it being operational with the current version of the MSL. Hence I **strongly** recommend to keep the version numbers in sync for all libraries shipped as part of the MSL package.

As for the ModelicaServices variants, misusing the `versionBuild` for this purpose is not one bit better than using a fourth digit or an additional string. Consider a tool ships MSL with ModelicaServices with its own flavour marked by a certain `versionBuild`. Now loading a user model that relies on that ModelicaServices variant will fail anyway no matter if you rely on the `versionBuild` that is not available in the other tool (but it will fail probably later at execution which is even more irritating since the user has no idea WHY) or if it fails because the special version of the ModelicaService variant is not avaiable (but here the user get's told why). 

The solution I see is to educate the tool vendors to NOT add any subversioning information when adding the `uses` annotation when it comes to ModelicaServices (i.e., only add `uses(ModelicaServices(version="3.2.1"))` but NOT `uses(ModelicaServices(version="3.2.1 xxx"))` . Still not ideal but less hacky I would say.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 2 Nov 2012 10:05 UTC**
e531e573f4eb7c64514ce211a9e8a9242aab4c2c also sets the version number of Complex in synch with MSL as argued above.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 2 Nov 2012 10:08 UTC**
What about ModelicaTest?

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 2 Nov 2012 10:26 UTC**
Replying to [comment:11 sjoelund.se]:
> What about ModelicaTest?

FIXED in 1ab228f349f14061d958b64aa0716aa1739ed9de
